File: app/string_similarity.rb
def self.z_function(string)
  length = string.length
  z_array = Array.new(length) { 0 }

  left, right, k = 0, 0, 1
  while k < length do
    if k > right
      left = right = k
      while right < length && string[right] == string[right - left] do
        right += 1
      end
      z_array[k] = right - left
      right -= 1
    else
      k1 = k - left
      if z_array[k1] < right - k + 1
        z_array[k] = z_array[k1]
      else
        left = k
        while right < length && string[right] == string[right - left] do
          right += 1
        end
        z_array[k] = right - left
        right -= 1 
      end
    end
    k += 1
  end

  z_array[0] = length
end

gets.to_i.times do
  s = gets.strip
  z_array = z_function(s)

  i = 0
  sum = 0
  while i < z_array.size do
    sum += z_array[i]
    i += 1
  end

  puts sum
end

File: app/string_similarity.go
package main

import (
  "fmt"
)

func main() {
  var t int
  var sum int
  var s string
  fmt.Scan(&t)

  for i := 0; i < t; i++ {
    fmt.Scan(&s)
    size := len(s)
    z_array := make([]int, size)
    z_array[0] = size
    left, right := 0, 0

    for k := 1; k < size; k++ {
      if k > right {
        left, right = k, k
        for ; right < size && s[right] == s[right - left]; {
          right += 1
        }
        z_array[k] = right - left
        right -= 1
      } else {
        k1 := k - left
        if z_array[k1] < right - k + 1 {
          z_array[k] = z_array[k1]
        } else {
          left = k
          for ; right < size && s[right] == s[right - left]; {
            right += 1
          }
          z_array[k] = right - left
          right -= 1
        }
      }
    }
    
    sum = 0
    for j := 0; j < size; j++ {
      sum += z_array[j]
    }

    fmt.Println(sum)
  }
}
